dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java

```diff
@@ -3474,6 +3487,258 @@ protected void checkLogsForErrorsAndWarnings(List<String> failuresRegExps, Strin
+            if (!servers.contains(FeatureReplacementAction.NO_SERVERS) &&
+                (servers.contains(serverName) || servers.contains(FeatureReplacementAction.ALL_SERVERS))) {
+
+                Set<String> expectedFeatures = new HashSet<>(featureReplacementAction.getAddFeatures()); //the expected features, if present
+                expectedFeatures.addAll(featureReplacementAction.getAlwaysAddFeatures());
+                Set<String> installedFeatures = getInstalledFeatures(); //the features actually installed at runtime
+
+                //expected feature -> actual feature
+                Map<String, String> unexpectedFeatures = new HashMap<>();
+
+                //compare each installed feature to the expected ones
+                for (String installedFeature : installedFeatures) {
+                    //ignore versionless features
+                    int dash = installedFeature.indexOf("-");
+                    if (dash > -1) {
+                        String versionlessInstalledFeature = removeFeatureVersion(installedFeature);
+                        for (String replacementFeature : expectedFeatures) {
+                            String versionlessReplacementFeature = removeFeatureVersion(replacementFeature);
+                            if (versionlessReplacementFeature.equalsIgnoreCase(versionlessInstalledFeature)) {
+                                //if the base feature name matches but the version does not, throw an exception.
+                                if (!replacementFeature.equalsIgnoreCase(installedFeature)) {
+                                    unexpectedFeatures.put(replacementFeature, installedFeature);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if (unexpectedFeatures.size() > 0) {
+                    String message = "Runtime features were not of the expected version for repeat action (Server: " + serverName + ", Action: " + action.getID() + ").\n";
+                    for (Map.Entry<String, String> entry : unexpectedFeatures.entrySet()) {
+                        message = message + "Expected: " + entry.getKey() + ", Actual: " + entry.getValue() + ".\n";
+                    }
+                    message = message
+                              + "This is usually caused by a feature not being explicitly set in the FAT's server.xml such that FeatureReplacementAction does not replace it properly.";
+
+                    //if this is a local run then always throw an exception
+                    //if not local then check if the server is exempt
+                    //if not exempt then throw exception, otherwise just output a message
+                    //check for exempt servers should eventually be removed
+                    if (REPEAT_FEATURE_CHECK_ERROR) {
+                        if (FAT_TEST_LOCALRUN) {
+                            throw new Exception(message);
```

Comment:
Do we need to alert the user running locally that they should remove the server from the exempt list?

Reply:
I supposed that would be the nice thing to do. However, if they see the issue locally, fix it and restest locally then being in the exempt list still will be a moot point (just untidy).

Reply:
Done